azurelinuxagent/common/cgroups.py

```diff
@@ -0,0 +1,698 @@
+
+        self.cgroups = {}
+
+        if not self.enabled():
+            return
+
+        system_hierarchies = os.listdir(BASE_CGROUPS)
+        for hierarchy in CGroups._hierarchies:
+            if hierarchy not in system_hierarchies:
+                raise CGroupsException("Hierarchy {0} is not mounted".format(hierarchy))
+
+            if self.is_wrapper_cgroup:
+                cgroup_path = path_maker(hierarchy)
+            else:
+                cgroup_path = os.path.join(path_maker(hierarchy), self.name)
+            if not os.path.exists(cgroup_path):
+                logger.info("Creating cgroup directory {0}".format(cgroup_path))
+                CGroups._try_mkdir(cgroup_path)
+            self.cgroups[hierarchy] = cgroup_path
+
+    @staticmethod
+    def is_systemd_manager():
+        """
+        Determine if systemd is managing system services. Many extensions are structured as a set of services,
+        including the agent itself; systemd expects those services to remain in the cgroups in which it placed them.
+        If this process (presumed to be the agent) is in a cgroup that looks like one created by systemd, we can
+        assume systemd is in use.
+
+        :return: True if systemd is managing system services
+        :rtype: Bool
+        """
+        if CGroups._use_systemd is None:
+            hierarchy = HIERARCHIES[0]
+            path = CGroups.get_my_cgroup_folder(hierarchy)
+            CGroups._use_systemd = path.startswith(CGroups.construct_systemd_path_for_hierarchy(hierarchy))
+        return CGroups._use_systemd
+
+    @staticmethod
+    def _try_mkdir(path):
+        """
+        Try to create a directory. If it already exists as such, do nothing. Raise appropriate exceptions if an error
+        should occur.
+
+        :param path: str
+        """
+        if not os.path.isdir(path):
+            try:
+                os.mkdir(path)
+            except OSError as e:
+                if e.errno == errno.EEXIST:
+                    if not os.path.isdir(path):
+                        raise CGroupsException(
+                            "Create directory for cgroup {0}: normal file already exists with that name".format(path)
+                        )
+                    else:
+                        pass    # There was a race to create the directory, but it's there now, and that's fine
+                elif e.errno == errno.EACCES:
+                    # This is unexpected, as the agent runs as root
+                    raise CGroupsException("Create directory for cgroup {0}: permission denied".format(path))
+                else:
+                    raise OSError(e)
+
+    @staticmethod
+    def construct_custom_path_for_hierarchy(hierarchy):
+        return os.path.join(BASE_CGROUPS, hierarchy, AGENT_NAME)
+
+    @staticmethod
+    def construct_systemd_path_for_hierarchy(hierarchy):
+        return os.path.join(BASE_CGROUPS, hierarchy, 'system.slice')
```

Comment:
Should system.slice be a value in the OS util class hierarchy?  Is is particular to a distro flavor, or is it always the same for systemd?

Reply:
Constant for systemd.